src/persist-client/examples/persistent_source.rs

```diff
@@ -0,0 +1,568 @@
+        ) -> TimestampBindings<Self::SourceTimestamp, Self::AssignedTimestamp>;
+    }
+
+    /// A set of timestamp bindings that was retrieved from a [`Timestamper`].
+    #[derive(Debug)]
+    pub struct TimestampBindings<ST: timely::PartialOrder + Debug, T: Debug> {
+        // sorted by source timestamp
+        pub bindings: Vec<(ST, T)>,
+        pub lower: Antichain<ST>,
+        pub upper: Antichain<ST>,
+    }
+
+    impl<ST: timely::PartialOrder + PartialOrd + Debug, T: Copy + Debug> TimestampBindings<ST, T> {
+        pub fn get_timestamp(&self, ts: ST) -> T {
+            assert!(
+                self.lower.less_equal(&ts),
+                "source timestamp {:?} is not beyond lowers: {:?}",
+                ts,
+                self.lower
+            );
+            assert!(!self.upper.less_than(&ts));
+
+            // Very naive solution for this!
+
+            let mut result_ts = None;
+
+            for (source_ts, assigned_ts) in self.bindings.iter() {
+                if &ts <= source_ts {
+                    result_ts = Some(assigned_ts);
+                } else {
+                    break;
+                }
+            }
+
+            match result_ts {
+                Some(binding) => *binding,
+                None => panic!("missing binding for {:?}, bindings: {:?}", ts, self),
+            }
+        }
+    }
+}
+
+mod impls {
+    use std::collections::hash_map::Entry;
+    use std::collections::HashMap;
+    use std::fmt::Debug;
+    use std::hash::Hash;
+    use std::sync::Arc;
+    use std::time::Duration;
+
+    use async_trait::async_trait;
+    use mz_ore::now::NowFn;
+    use timely::progress::Antichain;
+
+    use crate::api::{Batch, Source, TimestampBindings, Timestamper};
+
+    #[derive(Clone)]
+    pub struct SequentialSource {
+        count: usize,
+        wait_time: Duration,
+    }
+
+    impl SequentialSource {
+        /// Creates a new [`SequentialSource`] that waits for the given `wait_time` before each
+        /// batch is emitted.
+        pub fn new(wait_time: Duration) -> Self
+        where
+            Self: Sized,
+        {
+            Self {
+                count: 0,
+                wait_time,
+            }
+        }
+    }
+
+    #[async_trait]
+    impl Source for SequentialSource {
+        type SourceTimestamp = usize;
+        type D = usize;
+
+        async fn read_batch(
+            &mut self,
+            max_num_records: usize,
+        ) -> Option<Batch<Self::D, Self::SourceTimestamp>> {
+            let mut updates = Vec::new();
+
+            let lower = self.count;
+
+            for i in self.count..(self.count + max_num_records) {
+                updates.push((i, i));
+            }
+
+            self.count = self.count + max_num_records;
+
+            let batch = Batch {
+                updates,
+                lower: Antichain::from_elem(lower),
+                upper: Antichain::from_elem(self.count - 1),
```

Comment:
why the -1? upper is exclusive, right? also given that lower is set to self.count the next time through, don't the lowers+uppers of the two batches not line up?

Reply:
It's because I just wanted to get this to work without thinking too hard. The problem is that for determining timestamp bindings, we need an inclusive "upper", that is, just the highest offset in that batch (for Kafka).

And yes, this doesn't mesh well with how differential/timely use lower/upper. We might have to use separate terms/concepts here.

Reply:
Concretely, when a batch contains updates up to offset `t`, the upper (in differential world) would be `t+1`. But I'm only interested in minting a binding up to `t` because there probably is no data yet under `t+1`.

Also keep in mind that + and - might not always be easy to do on these "offset" types, _maybe_. Also, the examples use offsets, while Kafka would have `(partition, offset)`, but the ideas map 1-to-1.